### DISCUSSION TITLE
Issue with upgrading Kong custom plugin

### DISCUSSION THREAD

**Discussion by narendrapatel** [Kong/kong#7611] [Help] [answered]
Kong version 2.5.0
OS : Centos 7.9
POstgreSQL : 9.6

Trying to upgrade a custom plugin developed for Kong version 0.13.0. In this plugin we are extending the API endpoint to add custom functionality. Facing an error around it in the **before** section of the endpoint config.

**custom endpoint** : /eg-rate-limiting/rules/:rule_name

Errors in **Kong logs**: 2021/07/27 12:32:46 [error] 5004#0: *66 [kong] endpoints.lua:117 [postgres] **invalid primary key: '{id="not a primary key",rule_name="missing primary key"}'**, client: 192.168.33.1, server: kong_admin, request: "GET /eg-rate-limiting/rules/only_request_uri HTTP/1.1", host: "192.168.33.11:8001"

The said key **rule_name** is already a primary key in eg_ratelimiting_rules table in PostgreSQL.

schema in **daos.lua**:
```
local RULES_SCHEMA = {
	primary_key = {"rule_name"},
	name = "eg_ratelimiting_rules",
    fields = {
		{ rule_id = typedefs.uuid },
		{ rule_name = { type = "string" },},
		{ rule = { type = "string" }, }
    },
}
return {
  eg_ratelimiting_rules = RULES_SCHEMA
}
```

Endpoint config in **api.lua**:
```
local eg_ratelimiting_rules_schema = kong.db.eg_ratelimiting_rules.schema
["/eg-rate-limiting/rules/:eg_ratelimiting_rules"] = {
    schema = eg_ratelimiting_rules_schema,
    methods = {
      before = function(self, db, helpers) 
        local rules, err_t = endpoints.select_entity(self, db, eg_ratelimiting_rules_schema)
        if err_t then
          return endpoints.handle_error(err_t)
        end
        if not rules then
          return kong.response.exit(HTTP_NOT_FOUND, { message = "Not found" })
        end                                          
      self.rule = rules[1]
      end,
      GET  = endpoints.get_entity_endpoint(eg_ratelimiting_rules_schema),
  }
```

schema is **postgresql db**:
```
kong=> \d eg_ratelimiting_rules
Table "public.eg_ratelimiting_rules"
  Column   | Type | Modifiers 
-----------+------+-----------
 rule_id   | uuid | 
 rule_name | text | not null
 rule      | text | 
Indexes:
    "eg_ratelimiting_rules_pkey" PRIMARY KEY, btree (rule_name)

kong=> 
```
Any help would be appreciated.

**Comment by locao** [comment, marked as answer]
Hello @narendrapatel! 

It seems that Kong infers that your endpoint key is an id, which happens by your `daos.lua` is missing an `endpoint_key` field. Please take a look at how it is used in the [transformations test plugin](https://github.com/Kong/kong/blob/2.5.0/spec/fixtures/custom_plugins/kong/plugins/transformations/daos.lua). It might fix your issue.

**Reply by narendrapatel** [reply]
@locao
I confirm the ```endpoint_key``` config worked for me.
Thank you so much.
